src/blockchains/icp/icp_worker.ts

```diff
@@ -0,0 +1,211 @@
+    this.MAX_CONCURRENT_REQUESTS = settings.MAX_CONCURRENT_REQUESTS;
+    this.LOOP_INTERVAL_CURRENT_MODE_SEC = settings.LOOP_INTERVAL_CURRENT_MODE_SEC;
+    this.BURN_ADDRESS = settings.BURN_ADDRESS;
+    this.AUTH = Buffer.from(`${this.RPC_USERNAME}:${this.RPC_PASSWORD}`).toString('base64');
+  }
+
+  async init() {
+    this.lastConfirmedBlock = await this.getBlockNumber() - this.CONFIRMATIONS;
+  }
+
+  async getBlockNumber() {
+    try {
+        const response = await fetch(this.NODE_URL + '/network/status', {
+            method: 'POST',
+            headers: {
+                'Authorization': `Basic ${this.AUTH}`,
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({
+                network_identifier: {
+                    blockchain: 'Internet Computer',
+                    network: '00000000000000020101',
+                },
+                metadata: {},
+            }),
+        });
+
+        const data = await response.json();
+        const blockIndex = data.current_block_identifier?.index;
+        if (blockIndex !== undefined) {
+            return blockIndex;
+        } else {
+          logger.error('Block index not found in the response');
+        }
+    } catch (error) {
+      logger.error('Error fetching block index:', error);
+    }
+  }
+
+  async fetchBlock(block_index: number, retries = 3, retryDelay = 1000): Promise<ICPBlock> {
+    const fetchWithRetry = async (attempt: number): Promise<ICPBlock> => {
+      try {
+        const response = await fetch(this.NODE_URL + '/block', {
+          method: 'POST',
+          headers: {
+            'Authorization': `Basic ${this.AUTH}`,
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify({
+            network_identifier: {
+              blockchain: 'Internet Computer',
+              network: '00000000000000020101',
+            },
+            block_identifier: {
+              index: block_index,
+            },
+          }),
+        });
+
+        if (!response.ok) {
+          throw new Error(`Failed to fetch block: ${response.statusText}`);
+        }
+
+        const data = await response.json();
+        let block: ICPBlock = data.block;
+        return block;
+      } catch (error) {
+        if (attempt < retries) {
+          logger.info(`Attempt ${attempt + 1} failed. Retrying in ${retryDelay}ms...`);
+          await new Promise(res => setTimeout(res, retryDelay));
+          return fetchWithRetry(attempt + 1);
+        } else {
+          if (error instanceof Error) {
+            throw new Error(`Failed to fetch block after ${retries} retries: ${error.message}`);
+          } else {
+            throw new Error(`Failed to fetch block after ${retries} retries: Unknown error`);
+          }
+        }
+      }
+    };
+
+    return fetchWithRetry(0); // Start with the first attempt
+}
+
+  async getTransactions(blocks: ICPBlock[]) {
```

Comment:
This function holds all the domain specific knowledge, it should be unit tested.

Reply:
It is a pure function, so you can move it to a separate utils file and test it easily.